### PR TITLE
BREAKING CHANGE: Replace SwiftyJSON with Codable

### PR DESCRIPTION
**Pull request by Andrew-Lees11**
## Description
This pull requests removed the CouchDB dependency on SwiftyJSON. 

The API's to interact with the database have been replaced to use Codable objects. As such this is a breaking change. 

## Motivation and Context

This was requested to add type safety to the CouchDB API and was requested in issue #73.
This change removes the need to cast the response JSON to and from swift types. Instead you can just store and retrieve Codable objects.

## How Has This Been Tested?
The current tests have been updated to use the new API and were all passing when running locally.

## Checklist:
- [x] I have submitted a [CLA form](https://github.com/IBM-Swift/CLA)
- [x] If applicable, I have updated the documentation accordingly.
- [x] If applicable, I have added tests to cover my changes.

### CODE REVIEW

File: Sources/CouchDB/CouchDBClient.swift [hunk 1]
```diff
@@ -15,73 +15,54 @@
...
-        return UsersDatabase(connProperties: self.connProperties, dbName: "_users")
-    }
-
     /// Create a new `Database`.
     ///
     /// - parameters:
     ///     - dbName: String name of the database
-    ///     - callback: Callback containing the newly created `Database`, or an NSError on failure.
-    public func createDB(_ dbName: String, callback: @escaping (Database?, NSError?) -> ()) {
+    ///     - callback: Callback containing the newly created `Database`, or an `CouchDBError` on failure.
```
Comment (kilnerm) [inline Sources/CouchDB/CouchDBClient.swift]: This should be ", or a `CouchDBError`..."

File: Sources/CouchDB/CouchDBClient.swift [hunk 2]
```diff
@@ -90,21 +71,16 @@ public class CouchDBClient {
     ///
     /// - parameters:
     ///     - dbName: String name of the `Database` to look up.
-    ///     - callback: Callback containing the result of the lookup or an NSError if one occurred.
-    public func dbExists(_ dbName: String, callback: @escaping (Bool, NSError?) -> ()) {
+    ///     - callback: Callback containing the result of the lookup or an CouchDBError if one occurred.
```
Comment (kilnerm) [inline Sources/CouchDB/CouchDBClient.swift]: This should be "....or a CouchDBError..."

File: Sources/CouchDB/CouchDBClient.swift [hunk 3]
```diff
@@ -113,94 +89,44 @@ public class CouchDBClient {
     ///
     /// - parameters:
     ///     - database: An instance of the `Database` to delete.
-    ///     - callback: Callback containing an NSError if one occurred.
-    public func deleteDB(_ database: Database, callback: @escaping (NSError?) -> ()) {
+    ///     - callback: Callback containing an CouchDBError if one occurred.
```
Comment (kilnerm) [inline Sources/CouchDB/CouchDBClient.swift]: This should be "....containing a CouchDBError..."

File: Sources/CouchDB/CouchDBClient.swift [hunk 4]
```diff
@@ -113,94 +89,44 @@ public class CouchDBClient {
...
         deleteDB(database.name, callback: callback)
     }
 
     /// Delete a `Database` given its name.
     ///
     /// - parameters:
     ///     - dbName: String name of the `Database` to delete.
-    ///     - callback: Callback containing an NSError if one occurred.
-    public func deleteDB(_ dbName: String, callback: @escaping (NSError?) -> ()) {
+    ///     - callback: Callback containing an CouchDBError if one occurred.
```
Comment (kilnerm) [inline Sources/CouchDB/CouchDBClient.swift]: This should be "....containing a CouchDBError..."

File: Sources/CouchDB/CouchDBClient.swift [hunk 5]
```diff
@@ -113,94 +89,44 @@ public class CouchDBClient {
...
 
+    // MARK: UUID
+    
     /// Returns some UUIDs created by CouchDB.
     ///
     /// - parameters:
     ///     - count: The number of UUIDs to get.
-    ///     - callback: Callback containing an array of UUIDs or an NSError if one occured.
-    public func getUUIDs(count : UInt, callback : @escaping ([String]?, NSError?) -> Void) {
+    ///     - callback: Callback containing an array of UUIDs or an CouchDBError if one occured.
```
Comment (kilnerm) [inline Sources/CouchDB/CouchDBClient.swift]: This should be "....or a CouchDBError..."

File: Sources/CouchDB/CouchDBClient.swift [hunk 6]
```diff
@@ -113,94 +89,44 @@ public class CouchDBClient {
...
+            callback(uuids?["uuids"], nil)
         }
-        req.end()
     }
 
     /// Returns a UUID created by CouchDB.
     ///
-    /// - parameter callback: Callback containing the UUID or an NSError if one occurred.
-    public func getUUID(callback : @escaping (String?, NSError?) -> Void) {
+    /// - parameter callback: Callback containing the UUID or an CouchDBError if one occurred.
```
Comment (kilnerm) [inline Sources/CouchDB/CouchDBClient.swift]: This should be "....or a CouchDBError..."

File: Sources/CouchDB/ConnectionProperties.swift [hunk 7]
```diff
@@ -47,7 +45,7 @@ public struct ConnectionProperties {
     ///     - secured: Whether or not to use a secured connection.
     ///     - username: CouchDB admin username. Defaults to `nil`.
     ///     - password: CouchDB admin password. Defaults to `nil`.
-    public init(host: String, port: Int16, secured: Bool, username: String?=nil, password: String?=nil) {
+    public init(host: String, port: Int, secured: Bool, username: String?=nil, password: String?=nil) {
```
Comment (kilnerm) [inline Sources/CouchDB/ConnectionProperties.swift]: Given port is later cast to Int16 could we update the doc here to indicate a maximum value, or perhaps even prevent it being set above the max value?

File: Sources/CouchDB/CouchDBUtils.swift [hunk 8]
```diff
@@ -77,16 +41,27 @@ class CouchDBUtils {
...
-            //Log this exception
+            throw error
+        }
+    }
+    
+    class func getBodyAsError(_ response: ClientResponse) -> CouchDBError {
+        do {
+            var body = Data()
+            try response.readAllData(into: &body)
+            var error = try JSONDecoder().decode(CouchDBError.self, from: body)
```
Comment (kilnerm) [inline Sources/CouchDB/CouchDBUtils.swift:61]: A comment here explaining that error will be of type CouchDBError would be useful. I had to go look up JSONDecoder to work this out.

File: Sources/CouchDB/Database.swift [hunk 9]
```diff
@@ -30,86 +32,63 @@ public class Database {
...
-    /// Typealias used for building query parameters in view functions.
-    public typealias KeyType = AnyObject
-    #endif
-
     /// Query parameters for view functions from design documents.
     public enum QueryParameters {
         /// Includes conflicts information in response. Ignored if include_docs isn’t true. Default is false.
         case conflicts (Bool)
-
+
```
Comment (kilnerm) [inline Sources/CouchDB/Database.swift]: Would be good to clean up these white space changes.

File: Sources/CouchDB/Database.swift [hunk 10]
```diff
@@ -149,225 +130,112 @@ public class Database {
...
-        req.end()
+    ///     - document: The new `Document`.
+    ///     - callback: Callback containing the `DocumentResponse` or an CouchDBError.
+    public func create<D: Document>(_ document: D, callback: @escaping (DocumentResponse?, CouchDBError?) -> ()) {
+        let requestOptions = CouchDBUtils.prepareRequest(connProperties, method: "POST", path: "/\(escapedName)", hasBody: true)
+        CouchDBUtils.documentRequest(document: document, options: requestOptions, callback: callback)
     }
-
-    /// Retrieve all documents from the database
+
```
Comment (kilnerm) [inline Sources/CouchDB/Database.swift]: White space.

File: Sources/CouchDB/Database.swift [hunk 11]
```diff
@@ -149,225 +130,112 @@ public class Database {
...
     /// Executes the specified view function from the specified design document.
     ///
     /// - parameters:
     ///     - view: View function name String.
     ///     - design: Design document name.
     ///     - params: Query parameters for the function.
-    ///     - callback: Callback containing the JSON response or NSError if one occurred.
-    ///                 Refer to http://docs.couchdb.org/en/2.1.0/api/ddoc/views.html for info on JSON contents.
-    public func queryByView(_ view: String, ofDesign design: String, usingParameters params: [Database.QueryParameters], callback: @escaping (JSON?, NSError?) -> ()) {
+    ///     - callback: Callback containing either the `AllDatabaseDocuments` or an CouchDBError.
```
Comment (kilnerm) [inline Sources/CouchDB/Database.swift]: This should be "....or a CouchDBError..."

File: Sources/CouchDB/Database.swift [hunk 12]
```diff
@@ -506,24 +399,14 @@ public class Database {
     ///     - designName: Name String of the design document to delete.
     ///     - revision: The latest revision String of the design document to delete.
     ///     - failOnNotFound: Bool indicating whether to return an error if the design document was not found.
-    ///     - callback: Callback containing an NSError if one occurred.
-    public func deleteDesign(_ designName: String, revision: String, failOnNotFound: Bool = false, callback: @escaping (NSError?) -> ()) {
+    ///     - callback: Callback containing the `DocumentResponse` or an CouchDBError.
```
Comment (kilnerm) [inline Sources/CouchDB/Database.swift]: This should be "....or a CouchDBError..."

File: Sources/CouchDB/Database.swift [hunk 13]
```diff
@@ -532,51 +415,31 @@ public class Database {
     ///     - attachmentName: Attachment name String.
     ///     - attachmentData: The attachment Data.
     ///     - contentType: Attachment MIME type String.
-    ///     - callback: Callback containing the new revision String, the JSON response,
-    ///                 and an NSError if one occurred.
-    public func createAttachment(_ docId: String, docRevison: String, attachmentName: String, attachmentData: Data, contentType: String, callback: @escaping (String?, JSON?, NSError?) -> ()) {
-        var doc: JSON?
-        var revision: String?
+    ///     - callback: Callback containing the `DocumentResponse` or an CouchDBError.
```
Comment (kilnerm) [inline Sources/CouchDB/Database.swift]: This should be "....or a CouchDBError..."

File: Sources/CouchDB/Database.swift [hunk 14]
```diff
@@ -532,51 +415,31 @@ public class Database {
...
 
     /// Get an attachment associated with a specified document.
     ///
     /// - parameters:
     ///     - docId: Document ID String that the attachment is associated with.
     ///     - attachmentName: Name String for the desired attachment.
-    ///     - callback: Callback containing the retrieved attachment data, an NSError if one occurred,
-    ///                 and the content type of the attachment.
-    public func retrieveAttachment(_ docId: String, attachmentName: String, callback: @escaping (Data?, NSError?, String?) -> ()) {
+    ///     - callback: Callback containing either the retrieved attachment data and the content type of the attachment or an CouchDBError.
```
Comment (kilnerm) [inline Sources/CouchDB/Database.swift]: This should be "....or a CouchDBError..."

File: Sources/CouchDB/Database.swift [hunk 15]
```diff
@@ -587,22 +450,10 @@ public class Database {
     ///     - docId: Document ID String that the attachment is associated with.
     ///     - docRevision: Latest revision String of the document.
     ///     - attachmentName: Name String of the attachment to be deleted.
-    ///     - failOnNotFound: Bool indicating whether to return an NSError if the attachment could not be found.
-    ///     - callback: Callback containing an NSError if one occurred.
-    public func deleteAttachment(_ docId: String, docRevison: String, attachmentName: String, failOnNotFound: Bool = false, callback: @escaping (NSError?) -> ()) {
+    ///     - failOnNotFound: Bool indicating whether to return an CouchDBError if the attachment could not be found.
```
Comment (kilnerm) [inline Sources/CouchDB/Database.swift]: Wasn't this flag removed?

File: Sources/CouchDB/Database.swift [hunk 16]
```diff
@@ -587,22 +450,10 @@ public class Database {
...
+    ///     - callback: Callback containing either the `DocumentResponse` or an CouchDBError.
```
Comment (kilnerm) [inline Sources/CouchDB/Database.swift]: This should be "....or a CouchDBError..."

File: Tests/CouchDBTests/Utils.swift [hunk 17]
```diff
@@ -16,31 +16,25 @@
...
-#endif
-
 import Foundation
-import SwiftyJSON
+import CouchDB
 
 class Utils {
-
-    struct Credentials {
+
```
Comment (kilnerm) [inline Tests/CouchDBTests/Utils.swift]: White space

**Review by kilnerm — CHANGES_REQUESTED** [review]
Mostly grammar police and white space gripes.

Would be good if we can guard against overflowing the Int16 cast!

File: README.md [hunk 18]
```diff
@@ -86,13 +86,107 @@ The CouchDBSample executable demonstrates how to create, read, update and delete
...
 
-## Example
+## API Documentation
 
-For a more comprehensive example, you can follow the Kitura tutorial [Getting Started with Server-side Swift on raywenderlich.com](https://www.raywenderlich.com/180721/kitura-tutorial-getting-started-with-server-side-swift) that shows you how to create a backend API and then link this to a CouchDB instance running on your local machine.
+#### Document
+
+CouchDB is a NoSQL database for storing documents. A `Document` is any structure that can be represented as JSON and contains `_id` and `_rev` fields.  
+The `_id` field is the unique identifier for the document. If it is not set, a random UUID will be assigned for the document.  
+The `_rev` field is the revision of the document. It is returned when you make requests and is used to prevent conflicts from multiple users updating the same document.
```
Comment (helenmasters) [inline README.md]: Andy, can you bullet point or similar the definitions for _id field and _rev field (or put them in a separate paragraph)?

File: Sources/CouchDB/CouchDBUtils.swift (hunk 8 above)
Comment (kilnerm) [reply Sources/CouchDB/CouchDBUtils.swift:61]: Actually, ignore this!

**Review by kilnerm — APPROVED** [review]

File: Sources/CouchDB/CouchErrorResponse.swift [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+        if let id = id {
+            errorDescription += ", while processing: \(id)"
+        }
+        return errorDescription
+    }
+
+    /// The Document ID.
+    public let id: String?
+
+    /// The error that occured.
```
Comment (helenmasters) [inline Sources/CouchDB/CouchErrorResponse.swift]: occurred not occured

File: Sources/CouchDB/CouchErrorResponse.swift [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+    
+    init(_ code: Int, id: String? = nil, reason: String? = nil) {
+        self.statusCode = code
+        self.id = id
+        let error = HTTP.statusCodes[code] ?? String(code)
+        self.error = error
+        self.reason = reason ?? error
+    }
+    
+    // Don't decode statusCode, It is extracted from the HTTP request.
```
Comment (helenmasters) [inline Sources/CouchDB/CouchErrorResponse.swift]: It should have a lowercase i after the comma

File: Sources/CouchDBSample/main.swift [hunk 21]
```diff
@@ -91,128 +81,126 @@ let connProperties = ConnectionProperties(
...
-    username: username, // username
-    password: password  // password
+    username: nil,      // admin username
+    password: nil       // admin password
 )
 
-Log.info("Connection Properties:\n\(connProperties)")
+print("Connection Properties:\n\(connProperties)")
 
 // Create couchDBClient instance using conn properties
```
Comment (helenmasters) [inline Sources/CouchDBSample/main.swift]: conn properties should be either connProperties or probably better "connection properties" in this comment.

**Review by helenmasters — APPROVED** [review]
Approving as only 3 minor comments to be addressed.

**Pull request merged**